core/tests.py

```diff
@@ -0,0 +1,329 @@
+            'password2': 'bestaatookniet',
+            'email': 'kandi@kotkt.nl', 
+            'nickname': 'wijbestaanniet',
+        }
+        form = RegisterForm(data=form_data)
+
+        # Data that was entered is correct
+        self.assertTrue(form.is_valid())
+
+        # Ensure the correct object is returned (but not saved)
+        user = form.save(commit=False)
+        self.assertIsNotNone(user)
+        self.assertEquals(user.email, 'kandi@kotkt.nl')
+        self.assertTrue(user.check_password('bestaatookniet'))
+        self.assertEqual(user.first_name, 'wijbestaanniet')
+
+        user = User.objects.filter(username='schaduwkandi').first()
+        self.assertIsNone(user)
+
+        # Ensure the correct data is saved
+        form.save(commit=True)  
+        user = User.objects.filter(username='schaduwkandi').first()
+        self.assertIsNotNone(user)
+        self.assertEquals(user.email, 'kandi@kotkt.nl')
+        self.assertTrue(user.check_password('bestaatookniet'))
+        self.assertEqual(user.first_name, 'wijbestaanniet')
+
+    # Test if a registering fails if required fields are missing
+    def test_form_fields_missing(self):
+        form_data = {
+            'nickname': 'empty',       
+        }
+        form = RegisterForm(data=form_data)
+        
+        # Data that was entered is incorrect
+        self.assertFalse(form.is_valid())
+        
+        # Ensure that only one error was given per missing field
+        self.assertTrue(form.has_error('username'))
+        self.assertEqual(len(form.errors.as_data()['username']), 1)
+        self.assertTrue(form.has_error('password1'))
+        self.assertEqual(len(form.errors.as_data()['password1']), 1)
+        self.assertTrue(form.has_error('password2'))
+        self.assertEqual(len(form.errors.as_data()['password2']), 1)
+        self.assertTrue(form.has_error('email'))
+        self.assertEqual(len(form.errors.as_data()['email']), 1)
+        self.assertEqual(len(form.errors.as_data()), 4)
+
+    # Test if a registering fails if the two passwords do not match
+    def test_form_nonmatching_password(self):
+        form_data = {
+            'username': 'schaduwkandi',
+            'password1': 'bestaatookniet',
+            'password2': 'nomatch',
+            'email': 'kandi@kotkt.nl', 
+            'nickname': 'wijbestaanniet',
+        }
+        form = RegisterForm(data=form_data)
+        
+        # Data that was entered is incorrect
+        self.assertFalse(form.is_valid())
+        
+        # Ensure that only one error was given
+        self.assertTrue(form.has_error('password2'))
+        self.assertEqual(len(form.errors.as_data()['password2']), 1)
+   
+    # Test if a registering fails if username or email was already chosen by another user
+    def test_form_duplicate_field(self):
+        form_data = {
+            'username': 'schaduwbestuur',
+            'password1': 'bestaatookniet',
+            'password2': 'nomatch',
```

Comment:
Shouldn't the passwords match for this test?

Reply:
Yes they should, nicely spotted!